### PR TITLE
protocol/bc: add an Entry pointer to ValueSource, exclude it from hashing

### PR DESCRIPTION
**Pull request by bobg**
This change relieves the various Entry types that use `ValueSource` objects from each having to maintain its own separate `Entry` pointer (or list), and simplifies their construction. A struct tag is used to exclude the `ValueSource`'s embedded `Entry` from hashing.

A carved-off piece of https://github.com/chain/chain/pull/788

### CODE REVIEW

**Comment by bobg** [comment]
PTAL

File: protocol/bc/value.go [hunk 1]
```diff
@@ -4,4 +4,8 @@ type ValueSource struct {
 	Ref      Hash
 	Value    AssetAmount
 	Position uint64 // what int do we actually want?
+
+	// The Entry corresponding to Ref, if available
+	// The struct tag excludes the field from hashing
```
Comment (kr) [inline protocol/bc/value.go]: Nit: periods at the ends of the sentences?

File: protocol/bc/map.go [hunk 2]
```diff
@@ -124,18 +129,14 @@ func mapTx(tx *TxData) (headerID Hash, hdr *TxHeader, entryMap map[Hash]Entry, e
...
 
-	mux := NewMux(Program{VMVersion: 1, Code: []byte{0x51}}) // 0x51 == vm.OP_TRUE, minus a circular dependency on protocol/vm
-	for _, src := range muxSources {
-		// TODO(bobg): addSource will recompute the hash of
-		// entryMap[src.Ref], which is already available as src.Ref - fix
-		// this (and a number of other such places)
-		mux.addSource(entryMap[src.Ref], src.Value, src.Position)
-	}
-	_, err = addEntry(mux)
+	mux := NewMux(muxSources, Program{VMVersion: 1, Code: []byte{0x51}}) // 0x51 == vm.OP_TRUE, minus a circular dependency on protocol/vm
```
Comment (kr) [inline protocol/bc/map.go]: No more circular dependency as of #797?

File: protocol/bc/entry.go [hunk 3]
```diff
@@ -118,7 +118,13 @@ func writeForHash(w io.Writer, c interface{}) error {
 		return nil
 
 	case reflect.Struct:
-		for i := 0; i < v.NumField(); i++ {
+		typ := v.Type()
+		for i := 0; i < typ.NumField(); i++ {
+			sf := typ.Field(i)
+			if sf.Tag.Get("entry") == "-" {
```
Comment (kr) [inline protocol/bc/entry.go]: Nice!

File: protocol/bc/value.go [hunk 4]
```diff
@@ -4,4 +4,8 @@ type ValueSource struct {
 	Ref      Hash
 	Value    AssetAmount
 	Position uint64 // what int do we actually want?
+
+	// The Entry corresponding to Ref, if available
```
Comment (kr) [inline protocol/bc/value.go]: Are there any rules for when this is or isn't set? It would be comforting to know what to expect.

Comment (bobg) [inline protocol/bc/value.go]: Added a little commentary.

File: protocol/bc/entry.go (hunk 3 above)
Comment (bobg) [inline protocol/bc/entry.go]: How do you like "entry" as the tag?

**Comment by bobg** [comment]
Updated, PTAL

File: protocol/bc/entry.go (hunk 3 above)
Comment (kr) [inline protocol/bc/entry.go]: Seems perfect to me. Tags aren't so pervasive that I would be worried about [clashing with something else](https://github.com/golang/go/wiki/Well-known-struct-tags), and it's perfectly clear in context.

**Comment by kr** [comment]
LGTM

**Pull request merged**